pypoptools/pypoptesting/test_framework.py

```diff
@@ -0,0 +1,182 @@
+from enum import Enum
+from typing import Callable
+from node import Node
+
+import datetime
+import logging
+import os
+import subprocess
+import sys
+import tempfile
+import time
+
+
+class TestStatus(Enum):
+    PASSED = 1
+    FAILED = 2
+    SKIPPED = 3
+
+
+TEST_EXIT_PASSED = 0
+TEST_EXIT_FAILED = 1
+TEST_EXIT_SKIPPED = 77
+
+
+TMPDIR_PREFIX = "pypoptesting_"
+
+
+class SkipTest(Exception):
+    """This exception is raised to skip a test"""
+
+    def __init__(self, message):
+        self.message = message
+
+
+class PopIntegrationTestMetaClass(type):
+    """Metaclass for PopIntegrationTestFramework.
+    Ensures that any attempt to register a subclass of `PopIntegrationTestFramework`
+    adheres to a standard whereby the subclass overrides `set_test_params` and
+    `run_test` but DOES NOT override either `__init__` or `main`. If any of
+    those standards are violated, a ``TypeError`` is raised."""
+
+    def __new__(cls, clsname, bases, dct):
+        if not clsname == 'PopIntegrationTestFramework':
+            if not ('run_test' in dct and 'set_test_params' in dct and 'setup_network' in dct):
+                raise TypeError("PopIntegrationTestFramework subclasses must override "
+                                "'run_test', 'set_test_params' and 'setup_network'")
+            if '__init__' in dct or 'main' in dct:
+                raise TypeError("PopIntegrationTestFramework subclasses may not override "
+                                "'__init__' or 'main'")
+
+        return super().__new__(cls, clsname, bases, dct)
+
+
+class PopIntegrationTestFramework(metaclass=PopIntegrationTestMetaClass):
+    """Base class for a pop integration test script.
+    Individual pop integration test scripts should subclass this class and override the
+    set_test_params(), run_test() and setup_network() methods.
+    Individual tests can also override the following methods to customize the test setup:
+    - add_options()
+    - setup_chain()
+    - setup_nodes()
+    The __init__() and main() methods should not be overridden.
+    This class also contains various public and private helper methods."""
+
+    def __init__(self):
+        """Sets test framework defaults. Do not override this method. Instead, override the set_test_params() method"""
+        self.success = False
+        self.nodes = []
+        self.num_nodes = 0
+        self.set_test_params()
+
+    def main(self, factory_lambda: Callable[[], Node]):
```

Comment:
```suggestion
    def main(self, factory_lambda: Callable[[Path], Node]):
```
We should pass path to `/tmp/$dir/node0` in factory, so that altchain will be configured on top of that dir

Reply:
Path is from https://docs.python.org/3/library/pathlib.html